### ISSUE TITLE
Navigation tracking

### ISSUE THREAD

**Issue by draganescu** [WordPress/gutenberg#29102]
⛏️ ⚒️  work in progress ⛏️ ⚒️

## High
-   Update Navigation Block Toolbar #28856 #29012
-   Combine Menu Switcher with Add Menu#28863
-   Update Navigation block inspector#28864 #29012
-   Reconsider placeholder state for the Link block #28440 #28861
-   Improve responsiveness of the navigation screen #29100
    -   Block toolbar disappears on small screen sizes #28875
-   Show indication of in progress for deleting menus #25435
-   Save navigation not working when site is not in domain root #25181
-   Preload API requests #24642 #28701
-   Add link variations for custom post types and taxonomies #24814
-   "Manage locations" in Navigation screen only showing 10 items #28933
-   Improve the Manage locations default option value #25313
-   Test the Navigation screen extensibility #23739
    -   The `enqueue_block_editor_assets` action is not fired #21648
    -   It is not possible to add custom formats/buttons to menu item toolbar#21649
-   Do an accessibility audit of the new editor
## Normal
-   Block styles not loading when using an block based theme #28874
-   Decide if it's better to switch to a Legacy Menu block
-   Migrate to bulk REST API #24907
-   List View Design Updates #24029
    -   Try drill down for highly nested items in the navigator #23583
    -   Allow expanding and collapsing of nested blocks in List View #26141
-   Introduce the new Navigation editor to the world #24683
-   Fix styling of allowed blocks #29101
-   Mark invalid menu items in Navigation editor and block #23573
-   Reconsider placeholder state#23953
-   Page List block styling is inconsistent with navigation link styling #29101
-   Load last edited menu by default#22623
-   Introduce reorder menu endpoint #25093
## Low
-   Make the `block-nav-menus` setting more granular #24638
-   Add new block-based navigation panel to Customizer #24844
-   Improve how non-link blocks appear in old `nav-menus.php` screen #24639
-   Audit menus endpoint slashing #25094
-   Audit menus route filters. #25095

**Comment by draganescu** [comment]
Updated and deprioritized a couple of issues.

**Comment by javierarce** [comment]
I'm reviewing the navigation issues & enhancements, and here are some tickets that don't appear in the list above:

- #30006
- #31100
- #31182
- #31276
- #31667
- #31813
- #33271
- #33272

**Comment by getdave** [comment]
I've moved all [the items from @javierarce's list](https://github.com/WordPress/gutenberg/issues/29102#issuecomment-885584039) that are not `Type: Bug` into the most appropriate places I could find.

Note this Issue's tasks are going to be re-prioritised in the near future. I will update here regarding when and where that will happen.

**Comment by getdave** [comment]
## Update - 2021-08-11

* Work has continued on bug fixes and stability.
* Looking to define what is required to bring the Nav Editor out of "experimental".
* To this end, the team is looking to arrange a hallway hangout (or similarly public forum) to compare/contrast the classic navigation screen with the block-based navigation screen. 
* We're also looking to filter out or break down any "Epic" Issues into more actionable tasks (where possible).
* @javierarce has also been looking into and documenting the need to add some "polish" to the Nav Editor screen in order that UX bugs/issues don't spoil the overall experience. I believe he will look to add these as Bugs to the repo soon.
* From this we will then look to reprioritise the task list in this tracking Issue around achieving feature parity at a user interface level (developer APIs will follow).

**Comment by javierarce** [comment]
Here are the last three issues I've recently created related to the editor. The first one is the most important and probably key to have a nice experience creating menus.

- #34035
- #34036
- #34032

I've added them (plus others from my previous message) to the original post on top. The only issue I see that is not categorized is: 

- #30006

**Comment by getdave** [comment]
I've added / moved all of the visual items raised by @javierarce to the `High` priority section. These are the _minimum_ baseline visual inconsistencies that need to be resolved prior to the screen coming out of "experimental".

> The only issue I see that is not categorized is `#30006`.

This is [the issue regarding supporting future block types](https://github.com/WordPress/gutenberg/issues/30006). I'd say that's beyond the scope of what is required to get the Nav Editor out of "experimental". However I will add to one of the lower priority sections.

**Comment by getdave** [comment]
## Update - 2021-08-25

- The team have been focused on reforming the project around the goal of removing the "experimental" status of the feature within the Gutenberg _Plugin_ (only).
- Held a Hallway Hangout on Tuesday to compare and contrast Navigation screens and determine next steps.
- ~A summary will be posted shortly~ [A summary of the meeting and decisions made is available](https://make.wordpress.org/core/2021/08/25/hallway-hangout-summary-compare-and-contrast-the-navigation-screens/).
- Tracking issue is being updated to focus on the tasks required to remove "experimental".
- Work is continuing on exploring wider architectural issues such as how to best make use of the Navigation _block_ whilst retaining stability.
- We've continued with small UI/UX fixes.

**Comment by talldan** [comment]
## Update - 2021-09-01
- A plan has been formed for migrating from the customize API to the REST API in the navigation editor. The editor has been temporarily using the customize API for saving menus while some REST API design issues were resolved (see: https://github.com/WordPress/gutenberg/issues/25093)
- The navigation editor header has been refreshed to more closely match the styles used by other editors (see: https://github.com/WordPress/gutenberg/pull/34166).
- Some early discussions have taken place in the #feature-navigation-block-editor slack channel about extensibility in the navigation screen. Join the discussion if that's a topic you're interested in (see: https://wordpress.slack.com/archives/C01KDAZJMQ9/p1630160027012100).

### Smaller fixes and enhancements
- The navigation block is due to support a new block type, the 'Submenu' block, and work has been undertaken to ensure the navigation editor supports this (see: https://github.com/WordPress/gutenberg/pull/34281).
- The above change also fixes many of the block styling issues in the navigation editor.
- A fix for the entities being shown in menu names was merged (see https://github.com/WordPress/gutenberg/pull/34263)
- Loading of the navigation screen has been optimized by preloading API requests (see https://github.com/WordPress/gutenberg/pull/34364)

**Comment by kellychoffman** [comment]
I did a review and noted UX items that should be addressed before we can move this out of experimental phase: (Will link them all up to issues)

- [x] "Switch menu" location is out of context from the current menu you are editing. We have solved this similar issue in template editor with a dropdown next to the name of the template you are on - we should reuse this UX.
- [x] Once you add a menu, the state of the block shows placeholder bars and you are blocked as to what to do next. Ideally, we skip this step and go right to add Existing Page, Add All Pages, Start Empty. (This layout is also currently broken and needs a bit of a refresh.)
- [ ] “Add new page” label is unclear - the key word to this setting is "Automatically" and that is only shown in the descriptor text. Let's move it to the label so it reads: "Add new pages automatically" 
- [x] No undo button - Is there a way to undo changes?
- [x] When adding all links I have to know to "Convert to links". This should either be automatic or more obvious.
- [ ] Manage locations button placement - is confusing in context and should be separated from the task of applying this menu to certain locations provided by the theme. (Current implementation has it hidden behind a tab.)
- [ ] Adding pages, whether one at a time or several at once needs some further iteration https://github.com/WordPress/gutenberg/pull/33849
- [ ] How can I see how this looks? Or apply different layouts/styles? ('View theme styles' similar to Widgets could be an MVP of this)
- [ ] Usage of "Menu" vs. "Navigation" terms feel a bit confusing. Not clear why we are using them + where. 

cc @javierarce

**Comment by getdave** [comment]
Thanks @kellychoffman. Great to have some additional eyes on this ✨ 

For the purposes of ensuring the list of "Todo" items is actionable, it would be super helpful if we can:

* ensure all the Issues raised are specific / immediately actionable.
* dedupe them against existing Issues.
* add them to the appropriate place under the `High` section.

> that should be addressed before we can move this out of experimental phase

As I'm sure you're aware, it's very easy for this list of "priority" items to grow indefinitely so I'm hopeful we can prune these back to the items that are absolutely _essential_ for removing the "experimental" flag 🙏.

For context, the general agreement in the Hallway Hangout was that we would focus on items that are clearly and obviously _detrimental_ to the UX in terms of being able to create a basic menu. Optimisations and ideal flows will be addressed in follow up phases.

Open to discussion however - how do others feel?

Thanks again @kellychoffman 🙇‍♂️

**Comment by talldan** [comment]
A lot of these are I think covered already, but maybe some things are missing from the tracking issue:
> "Switch menu" location is out of context from the current menu you are editing. We have solved this similar issue in template editor with a dropdown next to the name of the template you are on - we should reuse this UX.

Covered in https://github.com/WordPress/gutenberg/issues/31241, but this isn't part of the tracking issue.

> Once you add a menu, the state of the block shows placeholder bars and you are blocked as to what to do next. Ideally, we skip this step and go right to add Existing Page, Add All Pages, Start Empty. (This layout is also currently broken and needs a bit of a refresh.)

Covered in https://github.com/WordPress/gutenberg/issues/23953, already high priority

>  “Add new page” label is unclear - the key word to this setting is "Automatically" and that is only shown in the descriptor text. Let's move it to the label so it reads: "Add new pages automatically"

Good feedback, lets address this. Could be a Good First Issue.

> No undo button - Is there a way to undo changes?

Also part of https://github.com/WordPress/gutenberg/issues/31241. This is a quick win, so I've gone ahead and made the change in https://github.com/WordPress/gutenberg/pull/34533

> When adding all links I have to know to "Convert to links". This should either be automatic or more obvious.

This 'Convert to links' UI will be removed, the current behavior is a bug introduced from the nav block. (https://github.com/WordPress/gutenberg/issues/23953, https://github.com/WordPress/gutenberg/issues/33999)

> Manage locations button placement - is confusing in context and should be separated from the task of applying this menu to certain locations provided by the theme. (Current implementation has it hidden behind a tab.)

There's some related discussion on https://github.com/WordPress/gutenberg/issues/33815, where I've mentioned this. Would be great to see some options for improving this.

> Adding pages, whether one at a time or several at once needs some further iteration Split control for URL and Text within Link UI #33849

Seems to already be in progress, but not part of the tracking issue. At the moment there are a few issues around adding items, so I do agree the process could be revisited, but not sure I agree it's a blocker for opening up the editor in the plugin, but definitely something to address before this lands in WordPress core.

> How can I see how this looks? Or apply different layouts/styles? ('View theme styles' similar to Widgets could be an MVP of this)

This isn't possible right now, it will most likely be a longer term goal as part of integrating the customizer. The widgets screen has the same issue, but was still launched, so I don't think this is a blocker to removing the experimental flag.
 
>  Usage of "Menu" vs. "Navigation" terms feel a bit confusing. Not clear why we are using them + where.

Agree, it'd be good to have an issue for this and some guidance. I think this has come about because the navigation block is being used to edit classic menus.

**Comment by Mamaduka** [comment]
The #34224 issue is listed in both High and Normal priority lists.

**Comment by javierarce** [comment]
Thanks for the review, @kellychoffman! 

Regarding "How can I see how this looks? Or apply different layouts/styles?", I'd like to point out this exploration I did some time ago: 

- #31074

**Comment by getdave** [comment]
## Update - 2021-09-08

* Everyone is focused on addressing the `High` priority issues on the Tracking Issue (see above).
* Work has continued on [the Submenu block](https://github.com/WordPress/gutenberg/pull/33775) and [preparing it ready for use in the Nav Editor](https://github.com/WordPress/gutenberg/pull/33775).
* The navigation editor header has gained more tools to better match other editors - props to @talldan for all his work here.
* Work is underway to add [a global inserter to the screen](https://github.com/WordPress/gutenberg/pull/34619).
* The creation flow for navigation menus was improved with a [new block placeholder](https://github.com/WordPress/gutenberg/pull/34568)
* [Improvements to List View drag and drop](https://github.com/WordPress/gutenberg/pull/34022/) should have beneficial knock-on effects for the Nav Editor - props to @gwwar for her work here.
* We could do with [the ability to (multi) select non-consecutive blocks](https://github.com/WordPress/gutenberg/issues/34232). [This PR from @swissspidy looks promising](https://github.com/WordPress/gutenberg/pull/16811) but additional help much appreciated.

**Comment by spacedmonkey** [comment]
I created a number of issues related to menus. 

- [Improve handling of html markup in Navigation Label](https://github.com/WordPress/gutenberg/issues/34399)
- ✅ [Add block inserter to navigation editor](https://github.com/WordPress/gutenberg/issues/34438)
- ✅ [Add support for custom post types in navigation editor](https://github.com/WordPress/gutenberg/issues/34440)
- ✅ [Add support for custom taxonomies in navigation editor](https://github.com/WordPress/gutenberg/issues/34442)
- ✅ [Navigation editor should respect show_in_nav_menus](https://github.com/WordPress/gutenberg/issues/34443)

Adding support from custom post types / taxonomies should be a blocker for getting this out of experiment, as it would make the menu screen unusable for those with CPT / CT.

**Comment by talldan** [comment]
> The #34224 issue is listed in both High and Normal priority lists.

#34232 was also in both, so I've removed it from high priority. I feel like we even said it was low priority in the hallway hangout, but can't remember for sure.

**Comment by getdave** [comment]
Thanks for creating these! 👍 

> I created a number of issues related to menus.

~@spacedmonkey Have you added these to the relevant sections above? If not then we'll need to look at that. I'll check now.~

**Update**: I marked ✅ next to [all your items](https://github.com/WordPress/gutenberg/issues/29102#issuecomment-915127526) which I've added to the relevant sections. One is remaining as I'm genuinely not sure about it and would like to discuss and learn more from you.

> Adding support from custom post types / taxonomies should be a blocker for getting this out of experiment, as it would make the menu screen unusable for those with CPT / CT.

Anything that makes the screen unusable feels like a blocker. Let's investigate this one further. Adding to `High` list.

**Comment by spacedmonkey** [comment]
I also work like to flag this issue https://github.com/WordPress/gutenberg/issues/31551 . It is important that we maintain backwards compatibility and solve backwards compat for ACF, we likely cover lots of other use cases of users that have extended the current menu screens. i have posted my idea of compatibility https://github.com/WordPress/gutenberg/issues/31551#issuecomment-915886641. I would love some feedback.

**Comment by talldan** [comment]
## Update - 2021-09-15

- Work is still progressing on migrating the navigation editor fully over to the REST API. The REST API changes have been merged, and now the front end code is being updated.
- Lots of UI changes this week with the editor top bar being updated, the main block inserter added, and fixes to the block's styles
- Good discussion underway on [how to provide backwards compatibility for custom fields on the existing Menus screen](https://github.com/WordPress/gutenberg/issues/31551).
- **Up next**: We're [looking into bulk adding links](https://github.com/WordPress/gutenberg/issues/31667), and [using theme.json to control aspects of the navigation block](https://github.com/WordPress/gutenberg/pull/34784).

We're always looking for contributors to help work on the Nav Editor and block - it you're interested please head over to the [#feature-navigation-block-editor Slack channel](https://wordpress.slack.com/archives/C01KDAZJMQ9). Look forward to seeing you there.

**Comment by getdave** [comment]
## Update - 2021-09-22

* Lots of working happening on `High` priority items. Thanks for everyone who is contributing so much work 🙌 
* REST API interactions have been improved and more issues identified.
* A tonne of bugs got fixed 🐛 
* Started looking at ways to add links in Bulk.
* [Using Theme JSON to control the block in the editor has been ruled out](https://github.com/WordPress/gutenberg/pull/34784#issuecomment-924063034) 😞 
* Another Hallway Hangout is on the cards - dates/times to be confirmed.

**Comment by getdave** [comment]
## Update - 2021-09-29

* Discussion is ongoing on the best way to ensure interoperability and compatibility between Nav block and Nav Editor.
* Proposal to [utilise a new classic Menu block](https://github.com/WordPress/gutenberg/pull/34496) in the Nav Editor is ongoing. Any feedback welcomed.
* Folks (particularly @spacedmonkey) have been [working hard on backwards compatibility](https://github.com/WordPress/gutenberg/issues/35157). Any more feedback on [this Issue](https://github.com/WordPress/gutenberg/issues/31551) would be a great help.
* Lots of bug fixes continue to roll in for the Editor. Great work by everyone involved.

**Comment by getdave** [comment]
## Update - 2021-10-06

* Discussion is still continuing on the best way to ensure interoperability and compatibility between Nav block and Nav Editor.
* Looks like there will be a Hallway Hangout on Wednesday 13th October (TBC). Agenda will be posted to Make Blog.
* @talldan posted [a nice summary of current state of things in Slack](https://wordpress.slack.com/archives/C01KDAZJMQ9/p1633512526250600). There's a lot of great energy in the team right now and we are looking to coalesce around a shared set of objectives.

**Comment by getdave** [comment]
## Update - 2021-10-20

* The Nav **_Editor_** will **not** be included in WordPress 5.9 - this is because:
    - changes to the block are required for the editor to be a success (see below).
    - we need to allow sufficient time to test the editor before any major release and allow for community feedback.
* Following the Hallway Hangout a [summary was posted](https://make.wordpress.org/core/2021/10/15/summary-navigation-editor-and-block-hallway-hangout/). The outcome was:
    - in the short term, contributor efforts will switch to the Nav _block_ in order to resolve some of the underlying architectural issues
    - specifically separating the navigation's presentation from its data in order to make navigations reusable. This serves both the Nav Editor project and the WordPress 5.9 release in general.
    - The Navigation Editor will ultimately focus on manipulating the _data_ of a navigation which is why the above work is a prerequisite for the project's success.
    - Work on the Nav Editor will resume after WordPress 5.9. We will continue to focus on backwards compatibility whilst looking ahead to the world of blocks.
    - It's unlikely that we will pursue a new "Classic Menu" block. Rather focusing on the Navigation block (or its data).
* To this end @talldan has [two important Nav block PRs which everyone is encouraged to test and provide feedback on](https://github.com/WordPress/gutenberg/issues/35521#issuecomment-947549374).

**Comment by draganescu** [comment]
I will close this tracking issue for the time being. The work continues on the navigation block.
